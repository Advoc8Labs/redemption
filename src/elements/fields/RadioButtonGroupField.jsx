import React, { useState } from 'react';
import ReactForm from 'react-bootstrap/Form';
import { bootstrapStyles } from '../styles';

function RadioButtonGroupField({
  element,
  applyStyles,
  fieldLabel,
  required = false,
  fieldVal = '',
  otherVal = '',
  onChange = () => {},
  onOtherChange = () => {},
  onClick = () => {},
  elementProps = {},
  children
}) {
  const servar = element.servar;
  const [otherSelect, setOtherSelect] = useState({});
  const otherChecked =
    (otherSelect[servar.key] || fieldVal) && fieldVal === otherVal;
  return (
    <div
      css={{ ...applyStyles.getTarget('fc'), position: 'relative' }}
      {...elementProps}
    >
      {fieldLabel}
      {servar.metadata.options.map((opt, i) => {
        return (
          <ReactForm.Check
            type='radio'
            id={`${servar.key}-${i}`}
            key={`${servar.key}-${i}`}
            label={opt}
            checked={fieldVal === opt}
            required={required}
            onChange={onChange}
            onClick={onClick}
            value={opt}
            style={{
              display: 'flex',
              alignItems: 'flex-start',
              marginBottom: '18px',
              padding: 0,
              lineHeight: 'normal'
            }}
            css={{
              'input[type="radio"]': {
                margin: '5px 10px 5px 0',
                position: 'static'
              }
            }}
          />
        );
      })}
      {servar.metadata.other && (
        <div
          style={{
            display: 'flex',
            alignItems: 'center',
            marginBottom: '18px'
          }}
        >
          <ReactForm.Check
            type='radio'
            id={`${servar.key}-`}
            key={`${servar.key}-`}
            label='Other'
            checked={otherChecked}
            onChange={(e) => {
              setOtherSelect({
                ...otherSelect,
                [servar.key]: true
              });
              onChange(e);
            }}
            onClick={onClick}
            value={otherVal || ''}
            style={{
              display: 'flex',
              alignItems: 'flex-start',
              padding: 0,
              lineHeight: 'normal'
            }}
            css={{
              'input[type="radio"]': {
                margin: '5px 10px 5px 0',
                position: 'static'
              }
            }}
          />
          <ReactForm.Control
            type='text'
            css={{
              marginLeft: '5px',
              ...bootstrapStyles,
              ...applyStyles.getTarget('field'),
              '&:focus': applyStyles.getTarget('active'),
              '&:hover': applyStyles.getTarget('hover')
            }}
            id={servar.key}
            value={otherVal || ''}
            onChange={onOtherChange}
            onClick={onClick}
            maxLength={servar.max_length}
            minLength={servar.min_length}
            required={otherChecked}
            disabled={!otherChecked}
          />
        </div>
      )}
      {children}
    </div>
  );
}

export default RadioButtonGroupField;
