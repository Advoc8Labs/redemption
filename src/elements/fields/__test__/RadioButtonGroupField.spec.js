import React from 'react';
import { create, act } from 'react-test-renderer';
import RadioButtonGroupField from '../RadioButtonGroupField';
import ApplyStyles from '../../styles';

describe('RadioButtonGroupField', () => {
  it('renders a Radio Button group', () => {
    // Arrange
    const props = {
      element: {
        id: 'fieldId',
        apply_styles: {},
        styles: {
          height: '50',
          height_unit: '%',
          width: '50',
          width_unit: 'px',
          background_color: '000000',
          uploader_padding_top: '0',
          uploader_padding_bottom: '0',
          uploader_padding_left: '0',
          uploader_padding_right: '0',
          cta_padding_top: '0',
          cta_padding_bottom: '0',
          cta_padding_left: '0',
          cta_padding_right: '0'
        },
        mobile_styles: {},
        servar: {
          key: 'key',
          name: 'Upload file...',
          metadata: {
            file_types: 'image/*',
            options: []
          },
          type: 'select'
        },
        properties: { icon: 'https://feathery.tech' }
      },
      required: true,
      onChange: jest.fn(),
      onClick: jest.fn()
    };
    props.applyStyles = new ApplyStyles(props.element, []);

    // Act
    let component;
    act(() => {
      component = create(<RadioButtonGroupField {...props} />);
    });

    // Assert
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a Radio Button group with metadata options', () => {
    // Arrange
    const props = {
      element: {
        id: 'fieldId',
        apply_styles: {},
        styles: {
          height: '50',
          height_unit: '%',
          width: '50',
          width_unit: 'px',
          background_color: '000000',
          uploader_padding_top: '0',
          uploader_padding_bottom: '0',
          uploader_padding_left: '0',
          uploader_padding_right: '0',
          cta_padding_top: '0',
          cta_padding_bottom: '0',
          cta_padding_left: '0',
          cta_padding_right: '0'
        },
        mobile_styles: {},
        servar: {
          key: 'key',
          name: 'Upload file...',
          metadata: {
            file_types: 'image/*',
            options: ['opt1']
          },
          type: 'select'
        },
        properties: { icon: 'https://feathery.tech' }
      },
      required: true,
      onChange: jest.fn(),
      onClick: jest.fn()
    };
    props.applyStyles = new ApplyStyles(props.element, []);

    // Act
    let component;
    act(() => {
      component = create(<RadioButtonGroupField {...props} />);
    });

    // Assert
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a Radio Button group with metadata options and others field enabled', () => {
    // Arrange
    const props = {
      element: {
        id: 'fieldId',
        apply_styles: {},
        styles: {
          height: '50',
          height_unit: '%',
          width: '50',
          width_unit: 'px',
          background_color: '000000',
          uploader_padding_top: '0',
          uploader_padding_bottom: '0',
          uploader_padding_left: '0',
          uploader_padding_right: '0',
          cta_padding_top: '0',
          cta_padding_bottom: '0',
          cta_padding_left: '0',
          cta_padding_right: '0'
        },
        mobile_styles: {},
        servar: {
          key: 'key',
          name: 'Upload file...',
          metadata: {
            file_types: 'image/*',
            options: ['opt1']
          },
          type: 'select',
          other: true
        },
        properties: { icon: 'https://feathery.tech' }
      },
      required: true,
      onChange: jest.fn(),
      onClick: jest.fn()
    };
    props.applyStyles = new ApplyStyles(props.element, []);

    // Act
    let component;
    act(() => {
      component = create(<RadioButtonGroupField {...props} />);
    });

    // Assert
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
