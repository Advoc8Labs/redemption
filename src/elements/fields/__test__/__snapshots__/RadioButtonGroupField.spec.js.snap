// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButtonGroupField renders a Radio Button group 1`] = `
<div
  className="css-bk9udp-RadioButtonGroupField"
/>
`;

exports[`RadioButtonGroupField renders a Radio Button group with metadata options 1`] = `
<div
  className="css-bk9udp-RadioButtonGroupField"
>
  <div
    className="css-gwhptf form-check"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "marginBottom": "5px",
      }
    }
  >
    <input
      checked={false}
      className="form-check-input"
      disabled={false}
      id="key-0"
      onChange={[MockFunction]}
      onClick={[MockFunction]}
      required={true}
      type="radio"
      value="opt1"
    />
    <label
      className="form-check-label"
      htmlFor="key-0"
      title=""
    >
      opt1
    </label>
  </div>
</div>
`;

exports[`RadioButtonGroupField renders a Radio Button group with metadata options and others field enabled 1`] = `
<div
  className="css-bk9udp-RadioButtonGroupField"
>
  <div
    className="css-gwhptf form-check"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "marginBottom": "5px",
      }
    }
  >
    <input
      checked={false}
      className="form-check-input"
      disabled={false}
      id="key-0"
      onChange={[MockFunction]}
      onClick={[MockFunction]}
      required={true}
      type="radio"
      value="opt1"
    />
    <label
      className="form-check-label"
      htmlFor="key-0"
      title=""
    >
      opt1
    </label>
  </div>
</div>
`;
